### ISSUE TITLE
node compat: Dynamic import leads to missing globals

### ISSUE THREAD

**Issue by marvinhagemeister** [denoland/deno#20028]
A dynamic import that is not statically analysable breaks npm detection which leads to node globals missing.

This was originally reported on the Fresh repo https://github.com/denoland/fresh/issues/1577 . In Fresh the `dev.ts` entry point calls a dynamic import where the specifier [is not statically analysable](https://github.com/denoland/fresh/blob/71fc1b8efcb3d3a51d9e8216eab32634b3b1051e/src/dev/mod.ts#L192):

```ts
export async function dev(base: string, entrypoint: string) {
  entrypoint = new URL(entrypoint, base).href;
  // ...
  await import(entrypoint);
}
```

Reproduction repo: https://github.com/marvinhagemeister/deno-npm-bug

## Steps to reproduce

1. Clone https://github.com/marvinhagemeister/deno-npm-bug
2. Run `deno run -A foo.ts`

**Comment by bartlomieju** [comment]
CC @lucacasonato

**Comment by bartlomieju** [comment]
The contents of the file are:
```
"use strict";
/**
 * Node.js implementation of browser `btoa` function.
 *
 * @packageDocumentation
 * @internal
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.base64Encode = void 0;
/**
 * @internal
 */
function base64Encode(str) {
    return Buffer.from(str).toString("base64");
}
exports.base64Encode = base64Encode;
//# sourceMappingURL=btoa.js.map⏎
```

I think this might be the limitation Luca mentioned, but I'm not sure.

**Comment by lucacasonato** [comment]
No, this doesn't look related. Please bisect if you think it's related to my globals PR. I think what is happening is that `internals.node.initialize()` is either not being called, or there is a race. Can you find that out?

I am about 98% confident this bug was already present when I implemented `npm:` support in Fresh. I reported it to @dsherret back then.

**Comment by bartlomieju** [comment]
Ah, thanks for the pointer, that's probably related to https://github.com/denoland/deno/issues/15826 then.

**Comment by loynoir** [comment]
Minimal reproduce.

OK when

```ts
==> foo.test.ts <==
import kleur from 'npm:kleur@3.0.0'

Deno.test('x', () => {
  console.log({ kleur })
  kleur.red('1x')
})

==> test.ts <==
await import('./foo.test.ts')
```

Error when

```ts
==> foo.test.ts <==
import kleur from 'npm:kleur@3.0.0'

Deno.test('x', () => {
  console.log({ kleur })
  kleur.red('1x')
})

==> test.ts <==
await import(`./foo.test.ts`)
```

**Comment by williamhorning** [comment]
I've been able to reproduce this this with both Deno 1.35.3 and 1.36.0 and found that the globals for Node packages are set right if you statically import those packages before they are dynamically imported. [williamhorning/deno-dynamic-import-globals](https://github.com/williamhorning/deno-dynamic-import-globals) has an example of this using undici.

**Comment by davideperozzi** [comment]
@williamhorning Thank you! I've been looking for a temporary fix. Importing **any npm module** before the dynamic import did the trick. So I'm basically doing `import 'npm:is-number'` (just using it, becaus it's small) to force the globals to load.

**Comment by bartlomieju** [comment]
FYI you should be able to work-around the problem by importing any built-in `node:` module, no need to import `npm:` libraries.
